File: kotlin/src/test/kotlin/species/CreatorSpeciesResolverTest.kt
package species

import org.junit.jupiter.api.DynamicTest.dynamicTest
import org.junit.jupiter.api.TestFactory
import kotlin.test.Test
import kotlin.test.assertContains
import kotlin.test.assertEquals
import kotlin.test.assertFalse

class CreatorSpeciesResolverTest {
    // A test case
    data class TC<T>(
        val does: Collection<String>,
        val doesnt: Collection<String>,
        val expected: T,
    )

    /**
     * - Most species
     *   - Hidden
     * - Other
     */
    private fun getBasicSpecies(): Species {
        val species = Species.Builder()
        val mostSpecies = species.getByNameCreatingMissing("Most species")
        val hidden = species.getByNameCreatingMissing("Hidden", true)
        mostSpecies.addChild(hidden)
        val other = species.getByNameCreatingMissing("Other")

        species.addRootSpecie(mostSpecies)
        species.addRootSpecie(other)

        return species.getResult()
    }

    @Test
    fun `Empty both does and doesnt returns empty set`() {
        val subject = CreatorSpeciesResolver(getBasicSpecies())

        val result = subject.resolveDoes(setOf(), setOf())
        assertEquals(0, result.size)
    }

    @Test
    fun `Empty does and unknown doesn't returns 'Most species' only`() {
        val subject = CreatorSpeciesResolver(getBasicSpecies())

        val result = subject.resolveDoes(setOf(), setOf("Some unusual specie"))
        assertEquals(setOf("Most species"), result)
    }

    @Test
    fun `Hidden species are not returned`() {
        val subject = CreatorSpeciesResolver(getBasicSpecies())

        val result = subject.resolveDoes(setOf("Most species"), setOf())

        assertContains(result, "Most species")
        assertFalse(result.contains("Hidden"))
    }

    @Test
    fun `Other species are simplified`() {
        val subject = CreatorSpeciesResolver(getBasicSpecies())

        val result = subject.resolveDoes(setOf("Some weird specie"), setOf())

        assertContains(result, "Other")
        assertFalse(result.contains("Some weird specie"))
    }

    /**
     * - Most species
     *   - A
     *     - B
     *       - C
     *         - D
     * - Other
     */
    private fun getGetOrderedDoesDoesntSpecies(): Species {
        val species = Species.Builder()
        val mostSpecies = species.getByNameCreatingMissing("Most species")
        val a = species.getByNameCreatingMissing("A")
        val b = species.getByNameCreatingMissing("B")
        val c = species.getByNameCreatingMissing("C")
        val d = species.getByNameCreatingMissing("D")
        mostSpecies.addChild(a)
        a.addChild(b)
        b.addChild(c)
        c.addChild(d)
        val other = species.getByNameCreatingMissing("Other")

        species.addRootSpecie(mostSpecies)
        species.addRootSpecie(other)

        return species.getResult()
    }

    @TestFactory
    fun getOrderedDoesDoesnt() = listOf(
        TC(listOf("A", "C"), listOf("B", "D"), "+A -B +C -D"),
        TC(listOf("C", "A"), listOf("D", "B"), "+A -B +C -D"),
        TC(listOf("B", "D"), listOf("A", "C"), "-A +B -C +D"),
        TC(listOf("D", "B"), listOf("C", "A"), "-A +B -C +D"),
    ).map { case ->
        dynamicTest(case.toString()) {
            val subject = CreatorSpeciesResolver(getGetOrderedDoesDoesntSpecies())

            val result = subject.getOrderedDoesDoesnt(case.does, case.doesnt)
            val strResult = result.map { (specie, does) -> (if (does) "+" else "-") + specie.name }.joinToString(" ")

            assertEquals(case.expected, strResult)
        }
    }

    /**
     * - Most species
     *   - Mammals
     *     - Canines
     *       - Dogs
     *         - Corgis
     *         - Dalmatians
     *       - Wolves
     *     - Deers
     *   - With antlers
     *     - Deers
     * - Other
     */
    private fun getResolveDoesSpecies(): Species {
        val species = Species.Builder()
        val mostSpecies = species.getByNameCreatingMissing("Most species")
        val mammals = species.getByNameCreatingMissing("Mammals")
        val withAntlers = species.getByNameCreatingMissing("With antlers")
        val canines = species.getByNameCreatingMissing("Canines")
        val dogs = species.getByNameCreatingMissing("Dogs")
        val corgis = species.getByNameCreatingMissing("Corgis")
        val dalmatians = species.getByNameCreatingMissing("Dalmatians")
        val wolves = species.getByNameCreatingMissing("Wolves")
        val deers = species.getByNameCreatingMissing("Deers")
        mostSpecies.addChild(mammals)
        mostSpecies.addChild(withAntlers)
        mammals.addChild(canines)
        canines.addChild(dogs)
        dogs.addChild(corgis)
        dogs.addChild(dalmatians)
        canines.addChild(wolves)
        mammals.addChild(deers)
        withAntlers.addChild(deers)
        val other = species.getByNameCreatingMissing("Other")

        species.addRootSpecie(mostSpecies)
        species.addRootSpecie(other)

        return species.getResult()
    }

    @TestFactory
    fun resolveDoes() = listOf(
        // @formatter:off
        TC(setOf(),                       setOf(),                          setOf()),
        TC(setOf("Mammals", "Corgis"),    setOf("Canines", "With antlers"), setOf("Mammals", "Corgis")),
        TC(setOf("Mammals"),              setOf("With antlers", "Dogs"),    setOf("Mammals", "Canines", "Wolves")),
        TC(setOf("Mammals", "Deers"),     setOf("With antlers", "Dogs"),    setOf("Mammals", "Canines", "Wolves", "Deers")),
        TC(setOf("Dogs", "With antlers"), setOf(""),                        setOf("With antlers", "Deers", "Dogs", "Corgis", "Dalmatians")),
        TC(setOf("Dogs", "With antlers"), setOf("Deers"),                   setOf("With antlers", "Dogs", "Corgis", "Dalmatians")),
        TC(setOf("Dogs", "Pancakes"),     setOf(""),                        setOf("Other", "Dogs", "Corgis", "Dalmatians")),
        TC(setOf("Dogs", "Other"),        setOf("Dalmatians"),              setOf("Other", "Dogs", "Corgis")),
        // @formatter:on
    ).map { case ->
        dynamicTest(case.toString()) {
            val subject = CreatorSpeciesResolver(getResolveDoesSpecies())

            val result = subject.resolveDoes(case.does, case.doesnt)

            assertEquals(case.expected, result)
        }
    }
}

File: tests/ByNamespace/Species/CreatorSpeciesResolverTest.php
<?php

declare(strict_types=1);

namespace App\Tests\ByNamespace\Species;

use App\Species\CreatorSpeciesResolver;
use App\Species\Hierarchy\MutableSpecies;
use App\Species\Hierarchy\Specie;
use App\Species\Hierarchy\Species;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\Attributes\Small;
use PHPUnit\Framework\TestCase;
use Veelkoov\Debris\Sets\StringSet;
use Veelkoov\Debris\Vecs\StringVec;

#[Small]
class CreatorSpeciesResolverTest extends TestCase
{
    /**
     * - Most species
     *   - Hidden
     * - Other.
     */
    private function getBasicSpecies(): Species
    {
        $species = new MutableSpecies();
        $mostSpecies = $species->getByNameCreatingMissing('Most species', false);
        $hidden = $species->getByNameCreatingMissing('Hidden', true);
        $mostSpecies->addChild($hidden);
        $other = $species->getByNameCreatingMissing('Other', false);

        $species->addRootSpecie($mostSpecies);
        $species->addRootSpecie($other);

        return $species;
    }

    public function testEmptyBothDoesAndDoesntReturnsEmptySet(): void
    {
        $subject = new CreatorSpeciesResolver($this->getBasicSpecies());

        $result = $subject->resolveDoes(new StringVec(), new StringVec());
        self::assertCount(0, $result);
    }

    public function testEmptyDoesAndUnknownDoesntReturnMostSpeciesOnly(): void
    {
        $subject = new CreatorSpeciesResolver($this->getBasicSpecies());

        $result = $subject->resolveDoes(new StringVec(), StringVec::of('Some unusual specie'));
        self::assertEqualsCanonicalizing(['Most species'], $result->getValuesArray());
    }

    public function testHiddenSpeciesAreNotReturned(): void
    {
        $subject = new CreatorSpeciesResolver($this->getBasicSpecies());

        $result = $subject->resolveDoes(StringVec::of('Most species'), new StringVec());

        self::assertTrue($result->contains('Most species'));
        self::assertFalse($result->contains('Hidden'));
    }

    public function testOtherSpeciesAreSimplified(): void
    {
        $subject = new CreatorSpeciesResolver($this->getBasicSpecies());

        $result = $subject->resolveDoes(StringVec::of('Some weird specie'), new StringVec());

        self::assertTrue($result->contains('Other'));
        self::assertFalse($result->contains('Some weird specie'));
    }

    /**
     * - Most species
     *   - A
     *     - B
     *       - C
     *         - D
     * - Other.
     */
    private function getGetOrderedDoesDoesntSpecies(): Species
    {
        $species = new MutableSpecies();

        $mostSpecies = $species->getByNameCreatingMissing('Most species', false);

        $a = $species->getByNameCreatingMissing('A', false);
        $b = $species->getByNameCreatingMissing('B', false);
        $c = $species->getByNameCreatingMissing('C', false);
        $d = $species->getByNameCreatingMissing('D', false);

        $mostSpecies->addChild($a);
        $a->addChild($b);
        $b->addChild($c);
        $c->addChild($d);

        $other = $species->getByNameCreatingMissing('Other', false);

        $species->addRootSpecie($mostSpecies);
        $species->addRootSpecie($other);

        return $species;
    }

    /**
     * @return list<array{StringVec, StringVec, string}>
     */
    public static function getOrderedDoesDoesntSpeciesDataProvider(): array
    {
        return [
            [StringVec::of('A', 'C'), StringVec::of('B', 'D'), '+A -B +C -D'],
            [StringVec::of('C', 'A'), StringVec::of('D', 'B'), '+A -B +C -D'],
            [StringVec::of('B', 'D'), StringVec::of('A', 'C'), '-A +B -C +D'],
            [StringVec::of('D', 'B'), StringVec::of('C', 'A'), '-A +B -C +D'],
        ];
    }

    #[DataProvider('getOrderedDoesDoesntSpeciesDataProvider')]
    public function testGetOrderedDoesDoesnt(StringVec $does, StringVec $doesnt, string $expected): void
    {
        $subject = new CreatorSpeciesResolver($this->getGetOrderedDoesDoesntSpecies());

        $result = $subject->getOrderedDoesDoesnt($does, $doesnt);
        $strResult = StringVec::mapFrom($result,
            static fn (bool $does, Specie $specie) => ($does ? '+' : '-').$specie->getName())->join(' ');

        self::assertSame($expected, $strResult);
    }

    /**
     * - Most species
     *   - Mammals
     *     - Canines
     *       - Dogs
     *         - Corgis
     *         - Dalmatians
     *       - Wolves
     *     - Deers
     *   - With antlers
     *     - Deers
     * - Other.
     */
    private function getResolveDoesSpecies(): Species
    {
        $species = new MutableSpecies();
        $mostSpecies = $species->getByNameCreatingMissing('Most species', false);
        $mammals = $species->getByNameCreatingMissing('Mammals', false);
        $withAntlers = $species->getByNameCreatingMissing('With antlers', false);
        $canines = $species->getByNameCreatingMissing('Canines', false);
        $dogs = $species->getByNameCreatingMissing('Dogs', false);
        $corgis = $species->getByNameCreatingMissing('Corgis', false);
        $dalmatians = $species->getByNameCreatingMissing('Dalmatians', false);
        $wolves = $species->getByNameCreatingMissing('Wolves', false);
        $deers = $species->getByNameCreatingMissing('Deers', false);
        $mostSpecies->addChild($mammals);
        $mostSpecies->addChild($withAntlers);
        $mammals->addChild($canines);
        $canines->addChild($dogs);
        $dogs->addChild($corgis);
        $dogs->addChild($dalmatians);
        $canines->addChild($wolves);
        $mammals->addChild($deers);
        $withAntlers->addChild($deers);
        $other = $species->getByNameCreatingMissing('Other', false);

        $species->addRootSpecie($mostSpecies);
        $species->addRootSpecie($other);

        return $species;
    }

    /**
     * @return list<array{StringVec, StringVec, StringVec}>
     */
    public static function resolveDoesDataProvider(): array
    {
        return [
            [new StringVec(),                       new StringVec(),                          new StringVec()],
            [StringVec::of('Mammals', 'Corgis'),    StringVec::of('Canines', 'With antlers'), StringVec::of('Mammals', 'Corgis')],
            [StringVec::of('Mammals'),              StringVec::of('With antlers', 'Dogs'),    StringVec::of('Mammals', 'Canines', 'Wolves')],
            [StringVec::of('Mammals', 'Deers'),     StringVec::of('With antlers', 'Dogs'),    StringVec::of('Mammals', 'Canines', 'Wolves', 'Deers')],
            [StringVec::of('Dogs', 'With antlers'), StringVec::of(''),                        StringVec::of('With antlers', 'Deers', 'Dogs', 'Corgis', 'Dalmatians')],
            [StringVec::of('Dogs', 'With antlers'), StringVec::of('Deers'),                   StringVec::of('With antlers', 'Dogs', 'Corgis', 'Dalmatians')],
            [StringVec::of('Dogs', 'Pancakes'),     StringVec::of(''),                        StringVec::of('Other', 'Dogs', 'Corgis', 'Dalmatians')],
            [StringVec::of('Dogs', 'Other'),        StringVec::of('Dalmatians'),              StringVec::of('Other', 'Dogs', 'Corgis')],
        ];
    }

    #[DataProvider('resolveDoesDataProvider')]
    public function testResolveDoes(StringVec $does, StringVec $doesnt, StringVec $expected): void
    {
        $subject = new CreatorSpeciesResolver($this->getResolveDoesSpecies());

        $result = $subject->resolveDoes($does, $doesnt);

        self::assertEqualsCanonicalizing($expected->getValuesArray(), $result->getValuesArray());
    }

    /**
     * @return list<array{StringSet, StringSet}>
     */
    public static function resolveForFiltersDataProvider(): array
    {
        return [
            [new StringSet(), new StringSet()],
            [StringSet::of('Deers'),        StringSet::of('Deers', 'With antlers', 'Most species', 'Mammals')],
            [StringSet::of('With antlers'), StringSet::of('With antlers', 'Most species')],
            [StringSet::of('Other'),        StringSet::of('Other')],
        ];
    }

    #[DataProvider('resolveForFiltersDataProvider')]
    public function testResolveForFilters(StringSet $speciesNames, StringSet $expected): void
    {
        $subject = new CreatorSpeciesResolver($this->getResolveDoesSpecies());

        $result = $subject->resolveForFilters($speciesNames);

        self::assertEqualsCanonicalizing($expected->getValuesArray(), $result->getValuesArray());
    }
}
